metadata/metadata.go

```diff
@@ -0,0 +1,123 @@
+package metadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"time"
+
+	"github.com/mackerelio/mackerel-agent/config"
+	"github.com/mackerelio/mackerel-agent/logging"
+)
+
+var logger = logging.GetLogger("metadata")
+
+// Generator generates metadata
+type Generator struct {
+	Name         string
+	Config       *config.MetadataPlugin
+	Tempfile     string
+	PrevMetadata interface{}
+}
+
+// Fetch invokes the command and returns the result
+func (g *Generator) Fetch() (interface{}, error) {
+	message, stderr, exitCode, err := g.Config.Run()
+
+	if err != nil {
+		logger.Warningf("Error occurred while executing a metadata plugin %q: %s", g.Name, err.Error())
+		return nil, err
+	}
+
+	if stderr != "" {
+		logger.Warningf("metadata plugin %q outputs stderr: %s", g.Name, stderr)
+		// metadata plugin can output message to stderr for debugging and json to stdout
+	}
+
+	if exitCode != 0 {
+		return nil, fmt.Errorf("exits with: %d", exitCode)
+	}
+
+	var metadata interface{}
+	if err := json.Unmarshal([]byte(message), &metadata); err != nil {
+		return nil, fmt.Errorf("outputs invalid JSON: %v", message)
+	}
+
+	return metadata, nil
+}
+
+// Differs returns whether the metadata has been changed or not
+func (g *Generator) Differs(metadata interface{}) bool {
+	if g.PrevMetadata == nil {
+		g.LoadFromFile()
+	}
+	return !reflect.DeepEqual(g.PrevMetadata, metadata)
+}
+
+// LoadFromFile loads the previous metadata from file
+func (g *Generator) LoadFromFile() {
+	data, err := ioutil.ReadFile(g.Tempfile)
+	if err != nil { // maybe initial state
+		return
+	}
+	var metadata interface{}
+	if err := json.Unmarshal(data, &metadata); err != nil {
+		logger.Warningf("metadata plugin %q detected a invalid json in temporary file: %s", g.Name, string(data))
+		// ignore errors, the file will be overwritten by Save()
+		return
+	}
+	g.PrevMetadata = metadata
+}
+
+// Save stores the metadata locally
+func (g *Generator) Save(metadata interface{}) error {
+	g.PrevMetadata = metadata
+	data, err := json.Marshal(metadata)
+	if err != nil {
+		return fmt.Errorf("failed to marshal the metadata to json: %v %s", metadata, err.Error())
```

Comment:
metadata may be  large, so I think it's better to place error message before metadata.